File: python/app/dsa/arrays/Contains_Duplicate.py
from typing import List

# Question : https://neetcode.io/problems/contains-duplicate/question

class Solution:

    # Approach 1: TC: O(n), SC: O(n)
    def hasDuplicate_A1(self, nums: List[int]) -> bool:
        mySet = set()
        for num in nums:
            if num in mySet:
                return True
            mySet.add(num)
        return False

    # Approach 2: TC: O(n), SC: O(n)
    def hasDuplicate_A2(self, nums: List[int]) -> bool:
        mySet = set()
        for num in nums:
            mySet.add(num)
        if len(mySet) != len(nums):
            return True
        return False  
    
if __name__ == "__main__":
    sol = Solution()
    print(sol.hasDuplicate_A1([1, 2, 3, 1]))  # True
    print(sol.hasDuplicate_A2([1, 2, 3, 4]))  # False   

File: java/src/main/dsa/arrays/Contains_Duplicate.java
package dsa.arrays;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

// Question : https://neetcode.io/problems/contains-duplicate/question

public class Contains_Duplicate {

    public static void main(String[] args) {
        Contains_Duplicate containsDuplicate = new Contains_Duplicate();
        System.out.println(containsDuplicate.hasDuplicate_A1(new int[]{1,2,3,3}));
        System.out.println(containsDuplicate.hasDuplicate_A2(new int[]{1,2,3,3}));
    }

    // #Approach:1
    // TC : O(N) , SC : O(N)
    public boolean hasDuplicate_A1(int[] nums) {
        Set<Integer> duplicateSet = new HashSet<>();
        return Arrays.stream(nums).anyMatch(val -> !duplicateSet.add(val));
    }

    // #Approach:2
    // TC : O(N) , SC : O(1)
    public boolean hasDuplicate_A2(int[] nums) {
        return Arrays.stream(nums).distinct().count() != nums.length;
    }
}
